crates/networking/p2p/rlpx/connection/server.rs

```diff
@@ -0,0 +1,859 @@
+    pub(crate) blockchain: Arc<Blockchain>,
+    pub(crate) capabilities: Vec<Capability>,
+    pub(crate) negotiated_eth_capability: Option<Capability>,
+    pub(crate) negotiated_snap_capability: Option<Capability>,
+    pub(crate) last_block_range_update_block: u64,
+    pub(crate) broadcasted_txs: HashSet<H256>,
+    pub(crate) requested_pooled_txs: HashMap<u64, NewPooledTransactionHashes>,
+    pub(crate) client_version: String,
+    //// Send end of the channel used to broadcast messages
+    //// to other connected peers, is ok to have it here,
+    //// since internally it's an Arc.
+    //// The ID is to ignore the message sent from the same task.
+    //// This is used both to send messages and to received broadcasted
+    //// messages from other connections (sent from other peers).
+    //// The receive end is instantiated after the handshake is completed
+    //// under `handle_peer`.
+    /// TODO: Improve this mechanism
+    /// See https://github.com/lambdaclass/ethrex/issues/3388
+    pub(crate) connection_broadcast_send: RLPxConnBroadcastSender,
+    pub(crate) table: Arc<Mutex<KademliaTable>>,
+    pub(crate) backend_channel: Option<Sender<Message>>,
+    pub(crate) inbound: bool,
+}
+
+#[derive(Clone)]
+pub enum InnerState {
+    Initiator(Initiator),
+    Receiver(Receiver),
+    Established(Established),
+}
+
+impl RLPxConnectionState {
+    pub fn new_as_receiver(context: P2PContext, peer_addr: SocketAddr, stream: TcpStream) -> Self {
+        Self(InnerState::Receiver(Receiver {
+            context,
+            peer_addr,
+            stream: Arc::new(stream),
+        }))
+    }
+
+    pub fn new_as_initiator(context: P2PContext, node: &Node) -> Self {
+        Self(InnerState::Initiator(Initiator {
+            context,
+            node: node.clone(),
+        }))
+    }
+}
+
+#[derive(Clone)]
+pub enum CallMessage {}
+
+#[derive(Clone)]
+#[allow(private_interfaces)]
+pub enum CastMessage {
+    PeerMessage(Message),
+    BackendMessage(Message),
+    SendPing,
+    SendNewPooledTxHashes,
+    BlockRangeUpdate,
+    BroadcastMessage(task::Id, Arc<Message>),
+}
+
+#[derive(Clone)]
+#[allow(private_interfaces)]
+pub enum OutMessage {
+    InitResponse {
+        node: Node,
+        framed: Arc<Mutex<Framed<TcpStream, RLPxCodec>>>,
+    },
+    Done,
+    Error,
+}
+
+#[derive(Debug)]
+pub struct RLPxConnection {}
+
+impl RLPxConnection {
+    pub async fn spawn_as_receiver(
+        context: P2PContext,
+        peer_addr: SocketAddr,
+        stream: TcpStream,
+    ) -> RLPxConnectionHandle {
+        let state = RLPxConnectionState::new_as_receiver(context, peer_addr, stream);
+        RLPxConnection::start(state)
+    }
+
+    pub async fn spawn_as_initiator(context: P2PContext, node: &Node) -> RLPxConnectionHandle {
+        let state = RLPxConnectionState::new_as_initiator(context, node);
+        RLPxConnection::start(state.clone())
+    }
+}
+
+impl GenServer for RLPxConnection {
+    type CallMsg = Unused;
```

Comment:
Nit: If we have unused here, do we need to define the empty call message struct above?

Reply:
Not really, good find. Don't know why rust forgot to complain about an unused struct.